clang/include/clang/Sema/Sema.h

```diff
@@ -14037,6 +14037,15 @@ class Sema final : public SemaBase {
       TemplateArgument Arg,
       SmallVectorImpl<UnexpandedParameterPack> &Unexpanded);
 
+  /// Collect the set of unexpanded parameter packs within the given
+  /// template argument.
+  ///
+  /// \param Arg The template argument that will be traversed to find
+  /// unexpanded parameter packs.
+  void collectUnexpandedParameterPacksForFoldExprs(
+      Expr *E, SmallVectorImpl<UnexpandedParameterPack> &Unexpanded,
+      SmallVectorImpl<UnexpandedParameterPack> &UnexpandedFromConstraints);
+
```

Comment:
I'm using a different API to collect *unexpanded constaints* - they have to be handled individually because we don't transform the constaint expressions like what we have done for other temploids, so typically we don't have sufficient template arguments for them during the transform of CXXFoldExprs. (See my comment https://github.com/llvm/llvm-project/pull/86265#discussion_r1693246465)

Reply:
What makes fold expressions different from any other pack expansion?
For example,

```cpp
void f(auto...)
template <class = void> void foo() {
 []<class... Is>() {
    f([]() requires (!C<Is>) {}()...);
  }.template operator()<char, int, float>();
}
```cpp

Isn't that the same scenario?

(and please check the comment, it does look like an incorrect copy paste!)

Reply:
> (and please check the comment, it does look like an incorrect copy paste!)
Of course, I'll clean up all the comments (some of them are pretty unclear) until we have an agreement :)

Reply:
> For example,
> ...
> Isn't that the same scenario?

(Funny enough, I tried locally and found that we'd never build up a CXXFoldExpr for your example and inevitably... we crash 😅)

Regardless, I'll use my example to demonstrate:

```cpp
template <class = void> void f() {
  []<class... Is>() {
    ([]<class = void>()
       requires (!C<Is>)
     {}(),
     ...);
  }.template operator()<char, int, float>();
}
```

The scenario here is a little subtle. So previously, our expansion for CXXFoldExpr was like:
```cpp
// 1. collect unexpanded parameter packs from the "pattern" so admittedly we would have an unexpanded pack for the constraint
getSema().collectUnexpandedParameterPacks();

// 2. Check whether we should expand with template arguments.
getDerived().TryExpandParameterPacks();

if (!Expand) {
   // 3. Hold off. Rebuild an unexpanded CXXFoldExpr
} else {
   // 4. Otherwise, go on and expand.
}
```

When we want to expand the expression, we're actually in the context of instantiating the body of `foo()`. So we're now given one level of template arguments which is `<void>`.

This is insufficient for the expansion of the inner lambda `[]() requires (!C<Is>) {}`, right? Therefore, we would rebuild the whole pattern, as part of the transform of a callee (`template operator()<char, int, float>()`). Then, we rebuild the call expression and evaluate the constraints there.

However, we would fail to evaluate the constraint because we haven't yet expand the inner pattern, because we don't have sufficient template arguments.

So I take an approach to figure out if we have only unexpanded constraint expressions left. If that is the case, we don't use the current template arguments; instead, we collect them like what we do for constraints: so here we want
two levels of template arguments: `<char, int, float>` and `<void>`.

Using a dedicated version of `collectUnexpandedParameterPacks` could make such this possible as we don't have a way to tell where the unexpanded pack comes from otherwise.

Reply:
Let me rephrase. What is the difference between a fold expression and the expansion of parameter pack in a function call or other sort of expressions that could try to expand an outer pack?
Doing something specific for fold expression sounds very wrong

So, I have a concrete suggestion here... can we deal with constraints in a separate PR?
It might help us make progress better.
WDYT?

Reply:
> So, I have a concrete suggestion here... can we deal with constraints in a separate PR?
> It might help us make progress better.
> WDYT?

Great. I find I'm getting stuck with it now. Some other cases show up as I'm trying to figure out another approach, so I probably need extra effort here.
Moving forward without the constraints sounds good to me.